Add Cluster UUID to Persisted Index Metadata

abeyad opened the pull request:
This PR adds the cluster UUID to the persisted index metadata for
the cluster state, and separates the representation of the index
metadata on disk from the index metadata that is passed around in
cluster state.

The cluster UUID on the index metadata is used to check whether a dangling
index on a node's disk should be imported into the cluster. If the
cluster UUID in the cluster state is the same as the cluster UUID on the
index metadata, then the dangling index is not imported (and instead,
cleanup happens on disk). If the cluster UUID is not the same as the
cluster UUID on the index metadata, we can assume the cluster
experienced a drastic restart, so we import the dangling index into the
cluster to prevent loss of data.

abeyad commented:
@bleskes @dakrone I still need to add tests and do more rigorous manual testing, but the basics work based on basic testing.  Wanted preliminary feedback on the approach.

bleskes commented on core/src/main/java/org/elasticsearch/cluster/metadata/IndexMetaData.java:
```diff
@@ -526,12 +561,21 @@ public IndexMetaData apply(IndexMetaData part) {
         builder.version(in.readLong());
         builder.state(State.fromId(in.readByte()));
         builder.settings(readSettingsFromStream(in));
+        if (doesClusterUUIDExistOnVersion(builder.settings)) {
```
we only use toXContent for file persistency. Over the network we can also read uuids..

bleskes commented on core/src/main/java/org/elasticsearch/gateway/DanglingIndicesState.java:
```diff
@@ -119,12 +120,31 @@ void findNewAndAddDanglingIndices(MetaData metaData) {
+                            // the index was deleted) or failed cleanup
+                            logger.info("[{}] dangling index exists on the local file system, but it is deleted in the metadata of the " +
+                                        "same cluster that created it, deleting the index on the local file system", indexName);
+                            nodeEnv.deleteIndexDirectoryUnderLock(indexMetaData.getIndex(), new IndexSettings(indexMetaData, settings));
```
I think we should make this an explicit handling in the dangling pipeline. Just import them into the map here and do the following:

```
    public void processDanglingIndices(MetaData metaData) {
        if (nodeEnv.hasNodeFile() == false) {
            return;
        }
        cleanupAllocatedDangledIndices(metaData);
        findNewAndAddDanglingIndices(metaData);
        cleanDeletedIndices(metaData); <-- this is where this logic will live, before allocation.
        allocateDanglingIndices();
    }
```

This will make it easier to test as well...

bleskes commented:
@abeyad thx! I left some suggestions

abeyad commented:
@bleskes thank you for the feedback!  I incorporated the changes and also added some tests, including serialization tests for pre cluster UUID vs post.

abeyad commented:
@bleskes An example scenario which this implementation doesn't address is that we could have the following:

1) We have 3 nodes on pre 5.0 versions: `A`, `B, and `C`.  `A` is the master node.
2) All 3 nodes have index `idx`.
3) `C` goes offline.
4) `A` and `B` get upgraded to 5.0 (where cluster UUIDs are now taken into account on index metadata).
5) We delete `idx`, which takes effect on the cluster state for `A` and `B`.
6) `C` comes back online.  

At this point, `C` can not tell the difference between a real deleted index and one that it should import as dangling (because `A` and `B` could have genuinely been wiped out, in which case we would want to import as dangling).  However, there doesn't seem to be a workaround from this issue unless all of the nodes have the cluster UUID stored on index metadata, which is the whole point of this commit in the first place.  So I don't think its something we need to address but its just a caveat.

bleskes commented:
Thx @abeyad . I think that is fine. The "feature" is only enabled once all data folders have been updated. I will review the code as soon as I can.

ywelsch commented:
My concern with this change is that it unnecessarily blows up the cluster state by duplicating information that is already contained in another part of the state. What needs to be done imho is to separate the representation of the persisted index metadata from the index metadata that is part of the cluster state. For that I suggest introducing a new class that wraps index metadata and the cluster UUID and use that one to persist / load index metadata on the nodes.

bleskes commented:
>  separate the representation of the persisted index metadata from the index metadata that is part of the cluster state

Make sense to me. We should see how much craft that adds.. Perhaps make a sub class of `MetaDataStateFormat<IndexMetaData>`

abeyad commented:
@bleskes @ywelsch I've updated the code to create a separate notion of persisted index metadata vs. index metadata that is passed around in the cluster state.  Your feedback is appreciated.

@areek FYI since you were just working in some of the same classes for the index folder name change.

bleskes commented on core/src/main/java/org/elasticsearch/cluster/metadata/PersistedIndexMetaData.java:
```diff
@@ -0,0 +1,185 @@
+/**
+ * Index metadata that is persisted on disk.
+ */
+final public class PersistedIndexMetaData implements FromXContentBuilder<PersistedIndexMetaData>, ToXContent {
```
can we call this `IndexStateMetaData` (ala `ShardStateMetaData`)?

bleskes commented on core/src/main/java/org/elasticsearch/cluster/metadata/PersistedIndexMetaData.java:
```diff
@@ -0,0 +1,185 @@
+
+    @Override
+    public PersistedIndexMetaData fromXContent(XContentParser parser, ParseFieldMatcher parseFieldMatcher) throws IOException {
+        if (parser.currentToken() == null) { // fresh parser? move to the first token
```
can we use (or at least try to use) the ObjectParser to do this?  Also all of these conditionals in the begining feel weird. Do we need them or not?

abeyad replied:
@bleskes Thanks for alerting me to the ObjectParser.  In looking at it though, it may be tricky because during the parsing of the x-content, I have to make a decision whether to parse the old format (just IndexMetaData) or the new format (IndexStateMetaData).

bleskes commented on core/src/main/java/org/elasticsearch/cluster/metadata/PersistedIndexMetaData.java:
```diff
@@ -0,0 +1,185 @@
+
+    @Override
+    public String toString() {
+        try {
```
Strings.toString is handy here.. Also, do we really need all the metadata in the tostring? maybe just a summary?

bleskes commented on core/src/main/java/org/elasticsearch/gateway/DanglingIndicesState.java:
```diff
@@ -53,7 +57,8 @@
 
-    private final Map<Index, IndexMetaData> danglingIndices = ConcurrentCollections.newConcurrentMap();
+    private final Map<Index, IndexMetaData> danglingIndices = new HashMap<>();
+    private final Map<Index, IndexMetaData> deletedIndices = new HashMap<>();
```
why do we this as state? we prune them after any scan right?

abeyad replied:
You are correct, one option is that I can change `void findNewAndAddDanglingIndices` to `Map<Index, IndexMetaData> findOnDiskIndices`.... the method would then only add to the `danglingIndices` member variable, and return the `deletedIndices` to be used in the `cleanDeletedIndices` method directly.  What do you think?  I kind of just did it this way to be consistent with the `danglingIndices`... its not the cleanest implementation either way... I mainly did it to avoid having to re-read the on disk indices twice (once to get the dangling, once to get the deleted).

bleskes commented on core/src/main/java/org/elasticsearch/gateway/DanglingIndicesState.java:
```diff
@@ -108,37 +124,77 @@ void cleanupAllocatedDangledIndices(MetaData metaData) {
      * to the currently tracked dangling indices.
      */
     void findNewAndAddDanglingIndices(MetaData metaData) {
```
we should probably rename this method to indicates it searches for dangling and deleted indices.

bleskes commented on core/src/main/java/org/elasticsearch/gateway/DanglingIndicesState.java:
```diff
@@ -108,37 +124,77 @@ void cleanupAllocatedDangledIndices(MetaData metaData) {
      * Finds new dangling indices by iterating over the indices and trying to find indices
      * that have state on disk, but are not part of the provided meta data, or not detected
-     * as dangled already.
+     * as dangled already.  The second tuple value contains dangling indices that should not
```
can we elaborate here a bit as to why those indices are safe to delete?

bleskes commented on core/src/main/java/org/elasticsearch/gateway/DanglingIndicesState.java:
```diff
@@ -108,37 +124,77 @@ void cleanupAllocatedDangledIndices(MetaData metaData) {
+                        // dangling files are probably due to remnants from deleted indices (e.g. the node was offline while
+                        // the index was deleted) or failed cleanup
+                        logger.info("[{}] dangling index exists on the local file system, but it is deleted in the metadata of the " +
+                                    "same cluster that created it, deleting the index on the local file system", index.getName());
```
can we log the full index object, includeing the index uuid? also it would be good to have the current cluster uuid and the one in the index meta data state.

bleskes commented on core/src/main/java/org/elasticsearch/gateway/DanglingIndicesState.java:
```diff
@@ -108,37 +124,77 @@ void cleanupAllocatedDangledIndices(MetaData metaData) {
+                                    "same cluster that created it, deleting the index on the local file system", index.getName());
+                        delIndices.put(index, persistedIndex.getIndexMetaData());
+                    } else {
+                        if (metaData.hasIndex(index.getName())) {
```
just use `else if` ?

bleskes commented on core/src/main/java/org/elasticsearch/gateway/DanglingIndicesState.java:
```diff
@@ -108,37 +124,77 @@ void cleanupAllocatedDangledIndices(MetaData metaData) {
+     * and we have decided that the indices should not be imported as dangling.
+     */
+    void cleanDeletedIndices() {
+        if (deletedIndices.isEmpty()) {
```
this is unneeded - we just iterate of the list...

bleskes commented on core/src/main/java/org/elasticsearch/gateway/DanglingIndicesState.java:
```diff
@@ -108,37 +124,77 @@ void cleanupAllocatedDangledIndices(MetaData metaData) {
+        if (deletedIndices.isEmpty()) {
+            return;
+        }
+        for (Iterator<Map.Entry<Index, IndexMetaData>> entries = deletedIndices.entrySet().iterator(); entries.hasNext();) {
```
can we just do for each?

abeyad replied:
Now I can because I no longer need the iterator to remove the element from `deletedIndices` (since its no longer a state variable in the class).

bleskes commented on core/src/main/java/org/elasticsearch/gateway/DanglingIndicesState.java:
```diff
@@ -108,37 +124,77 @@ void cleanupAllocatedDangledIndices(MetaData metaData) {
+            final String indexName = entry.getKey().getName();
+            final IndexMetaData indexMetaData = entry.getValue();
+            try {
+                nodeEnv.deleteIndexDirectoryUnderLock(indexMetaData.getIndex(), new IndexSettings(indexMetaData, settings));
```
Oh oh! deleteUnderLock should be called when you hold the lock! instead we should use IndicesService.deleteIndexStore

bleskes commented on core/src/main/java/org/elasticsearch/gateway/DanglingIndicesState.java:
```diff
@@ -108,37 +124,77 @@ void cleanupAllocatedDangledIndices(MetaData metaData) {
+            try {
+                nodeEnv.deleteIndexDirectoryUnderLock(indexMetaData.getIndex(), new IndexSettings(indexMetaData, settings));
+                entries.remove();
+                logger.info("[{}] dangling index deleted from local file system", indexName);
```
I think this will result in a double info logging - we already logged at info level when discovering this.

bleskes commented on core/src/main/java/org/elasticsearch/gateway/GatewayMetaState.java:
```diff
@@ -223,19 +224,21 @@ private void ensureNoPre019State() throws Exception {
-        MetaData metaData = loadMetaState();
-        List<IndexMetaData> updateIndexMetaData = new ArrayList<>();
-        for (IndexMetaData indexMetaData : metaData) {
+        List<PersistedIndexMetaData> metaData = metaStateService.loadIndicesStatesNoFolderNameValidation();
```
why do we need the new loadIndicesStatesNoFolderNameValidation ? what was wrong with loadMetaState() exactly?

abeyad replied:
The `loadMetaState` method does not load `IndexStateMetaData` (the persisted class) but rather just `IndexMetaData`, which makes sense because it is used to load the `MetaData` object for the cluster state.  I could have used `loadIndicesStates`, but this method assumes that the index folder name change upgrade has executed that @areek implemented (it throws an exception if the folder path != index uuid).  I needed a method that skips this check as this code is executed before the index folder upgrade.

areek replied:
We can use `GatewayMetaState#loadMetaState` here, it uses `MetaStateService#loadFullState`, which loads the global state along with all the `IndexMetaData`. We can just iterate on the loaded index meta data and upgrade? we can use the `clusterUUID` from the global state here, since we are just upgrading or make the `MetaData.Builder` aware of `PersistedIndexMetaData` (don't know if this is worth it?)

bleskes commented on core/src/test/java/org/elasticsearch/cluster/metadata/PersistedIndexMetaDataTests.java:
```diff
@@ -0,0 +1,62 @@
+ */
+public class PersistedIndexMetaDataTests extends ESTestCase {
+
+    public void testSerialization() throws IOException {
```
can we make sure we read legacy files?

bleskes commented on core/src/test/java/org/elasticsearch/gateway/GatewayIndexStateIT.java:
```diff
@@ -364,4 +378,64 @@ public Settings onNodeStopped(String nodeName) throws Exception {
+        final AtomicBoolean failure = new AtomicBoolean(false);
+        final CountDownLatch latch = new CountDownLatch(1);
+        final CountDownLatch processedLatch = new CountDownLatch(1);
+        clusterService.submitStateUpdateTask("test", new AckedClusterStateUpdateTask<Void>(null, null) {
```
why not use a simple cluster state update task?

bleskes commented:
Left a bunch of comments. Thx @abeyad

abeyad commented on core/src/main/java/org/elasticsearch/gateway/DanglingIndicesState.java:
```diff
@@ -104,41 +116,85 @@ void cleanupAllocatedDangledIndices(MetaData metaData) {
+            final String indexName = entry.getKey().getName();
+            final IndexMetaData indexMetaData = entry.getValue();
+            try {
+                indicesService.deleteIndexStore("index deleted but metadata still on file system", indexMetaData, clusterState, false);
```
@bleskes I used the `IndicesService#deleteIndexStore` here, but it requires injection of the IndicesService and the cluster state, which means the `processDanglingIndices` method requires passing in `ClusterState` instead of just `MetaData`.  What do you think?

areek replied:
I think we should use:
```java
nodeEnv.deleteIndexDirectorySafe(indexMetaData.getIndex(), 0, new IndexSettings(indexMetaData, settings));
```
`indicesService.deleteIndexStore` additionally checks the cluster metadata to make sure we can delete the index store, IMO this check doesn't make sense here as we have already read the index in question as dangled indices w.r.t. the cluster metadata?

bleskes replied:
I know we already checked it but I think it's important to make sure all data deletion goes through the same place where we do (and will extend) all the required safety checks. We might even want to restrict access to the delete methods in NodeEnv via forbidden API.

abeyad commented:
@bleskes Thank you for reviewing the code.  I updated the PR to implement your comments, and left questions as well.

areek commented on core/src/main/java/org/elasticsearch/gateway/GatewayMetaState.java:
```diff
@@ -223,19 +224,21 @@ private void ensureNoPre019State() throws Exception {
-        MetaData metaData = loadMetaState();
-        List<IndexMetaData> updateIndexMetaData = new ArrayList<>();
-        for (IndexMetaData indexMetaData : metaData) {
+        List<IndexStateMetaData> metaData = metaStateService.loadIndicesStatesNoFolderNameValidation();
```
We can use `loadMetaState` here, it uses `MetaStateService#loadFullState`, which loads the global state along with all the `IndexMetaData`. We can just iterate on the loaded index meta data and upgrade? we can use the `clusterUUID` from the global state here, since we are just upgrading or make the `MetaData.Builder` aware of `PersistedIndexMetaData` (don't know if this is worth it?)

areek commented on core/src/main/java/org/elasticsearch/gateway/DanglingIndicesState.java:
```diff
@@ -104,41 +116,85 @@ void cleanupAllocatedDangledIndices(MetaData metaData) {
-    void findNewAndAddDanglingIndices(MetaData metaData) {
-        danglingIndices.putAll(findNewDanglingIndices(metaData));
+    Map<Index, IndexMetaData> findOnDiskAndAddDanglingIndices(MetaData metaData) {
+        Tuple<Map<Index, IndexMetaData>, Map<Index, IndexMetaData>> result = findDanglingAndDeletedIndices(metaData);
```
I think we can avoid returning tuples of maps here? I took a [stab](https://gist.github.com/areek/f7ebb27ab0b79d0bb422) at separating out logic for loading dangled index states and the logic for treating them as dangled/delete indices. WDYT?

areek commented on core/src/main/java/org/elasticsearch/cluster/metadata/IndexStateMetaData.java:
```diff
@@ -0,0 +1,170 @@
+    /**
+     * StateFormat that can read and write {@link IndexStateMetaData}
+     */
+    public static MetaDataStateFormat<IndexStateMetaData> FORMAT =
```
we should remove `IndexMetaData#FORMAT`and `IndexMetaData#FORMAT_PARAMS` because they are now unused because of this

abeyad commented:
Closing this PR without merging, because upon discussion with @bleskes, the change can cause us to lose indices in the case of split brain.  For example, suppose we have a split brain with master `A` on one side and master `B` on the other side.  If we create an index on `B` (it won't exist on `A`), then recover from split brain but `A`'s cluster state is the most recently updated so it becomes the master, then the index does not exist on `A`'s cluster state so it will delete the dangling indices instead of importing them.  

So, while the change itself is correct in terms of behavior, it can cause data loss in the split brain scenario, so the PR is being closed.  Another solution has been proposed where the cluster state will contain a list of tombstoned indices.  This will make deleted indices explicit in the cluster state instead of just relying on diffs in index metadata in the cluster state.
